File: csharp/src/Vista.SDK.System.Text.Json/DataChannelList/Extensions.cs
using Domain = Vista.SDK.Transport.DataChannel;

namespace Vista.SDK.Transport.Json.DataChannel;

public static class Extensions
{
    public static DataChannelListPackage ToJsonDto(this Domain.DataChannelListPackage package)
    {
        return new DataChannelListPackage(
            new Package(
                new DataChannelList(
                    package
                        .DataChannelList
                        .Select(
                            c =>
                                new DataChannel(
                                    new DataChannelID(
                                        c.DataChannelId.LocalId.VerboseMode
                                            ? c.DataChannelId.LocalId.Builder.WithVerboseMode(false).Build().ToString()
                                            : c.DataChannelId.LocalId.Builder.ToString(),
                                        c.DataChannelId.NameObject is null
                                            ? null
                                            : new NameObject(c.DataChannelId.NameObject.NamingRule)
                                            {
                                                CustomProperties = c.DataChannelId
                                                    .NameObject
                                                    .CustomNameObjects
                                                    ?.CopyProperties(),
                                            },
                                        c.DataChannelId.ShortId
                                    ),
                                    new Property(
                                        c.Property.AlertPriority,
                                        new DataChannelType(
                                            c.Property.DataChannelType.CalculationPeriod,
                                            c.Property.DataChannelType.Type,
                                            c.Property.DataChannelType.UpdateCycle
                                        ),
                                        new Format(
                                            c.Property.Format.Restriction is null
                                                ? null
                                                : new Restriction(
                                                    c.Property.Format.Restriction.Enumeration,
                                                    (int?)c.Property.Format.Restriction.FractionDigits,
                                                    (int?)c.Property.Format.Restriction.Length,
                                                    c.Property.Format.Restriction.MaxExclusive,
                                                    c.Property.Format.Restriction.MaxInclusive,
                                                    (int?)c.Property.Format.Restriction.MaxLength,
                                                    c.Property.Format.Restriction.MinExclusive,
                                                    c.Property.Format.Restriction.MinInclusive,
                                                    (int?)c.Property.Format.Restriction.MinLength,
                                                    c.Property.Format.Restriction.Pattern,
                                                    (int?)c.Property.Format.Restriction.TotalDigits,
                                                    (RestrictionWhiteSpace?)c.Property.Format.Restriction.WhiteSpace
                                                ),
                                            c.Property.Format.Type
                                        ),
                                        c.Property.Name,
                                        c.Property.QualityCoding,
                                        c.Property.Range is null
                                            ? null
                                            : new Range((double)c.Property.Range.High, c.Property.Range.Low),
                                        c.Property.Remarks,
                                        c.Property.Unit is null
                                            ? null
                                            : new Unit(c.Property.Unit.QuantityName, c.Property.Unit.UnitSymbol)
                                            {
                                                CustomProperties = c.Property.Unit.CustomElements?.CopyProperties()
                                            }
                                    )
                                    {
                                        CustomProperties = c.Property.CustomProperties?.CopyProperties()
                                    }
                                )
                        )
                        .ToArray()
                ),
                new Header(
                    package.Package.Header.Author,
                    new ConfigurationReference(
                        package.Package.Header.DataChannelListId.Id,
                        package.Package.Header.DataChannelListId.TimeStamp,
                        package.Package.Header.DataChannelListId.Version
                    ),
                    package.Package.Header.DateCreated,
                    package.Package.Header.ShipId.ToString(),
                    package.Package.Header.VersionInformation is null
                        ? null
                        : new VersionInformation(
                            package.Package.Header.VersionInformation.NamingRule,
                            package.Package.Header.VersionInformation.NamingSchemeVersion,
                            package.Package.Header.VersionInformation.ReferenceUrl
                        )
                )
                {
                    CustomHeaders = package.Package.Header.CustomHeaders?.CopyProperties()
                }
            )
        );
    }

    public static Domain.DataChannelListPackage ToDomainModel(this DataChannelListPackage package)
    {
        var p = package.Package;
        return new Domain.DataChannelListPackage
        {
            Package = new Domain.Package
            {
                Header = new Domain.Header
                {
                    ShipId = ShipId.Parse(p.Header.ShipID),
                    DataChannelListId = new Domain.ConfigurationReference
                    {
                        Id = p.Header.DataChannelListID.ID,
                        Version = p.Header.DataChannelListID.Version,
                        TimeStamp = p.Header.DataChannelListID.TimeStamp
                    },
                    VersionInformation = p.Header.VersionInformation is null
                        ? null
                        : new Domain.VersionInformation
                        {
                            NamingRule = p.Header.VersionInformation.NamingRule,
                            NamingSchemeVersion = p.Header.VersionInformation.NamingSchemeVersion,
                            ReferenceUrl = p.Header.VersionInformation.ReferenceURL
                        },
                    Author = p.Header.Author,
                    DateCreated = p.Header.DateCreated,
                    CustomHeaders = p.Header.CustomHeaders?.CopyProperties()
                },
                DataChannelList = new Domain.DataChannelList
                {
                    p.DataChannelList
                        .DataChannel
                        .Select(
                            c =>
                                new Domain.DataChannel
                                {
                                    DataChannelId = new Domain.DataChannelId
                                    {
                                        LocalId = LocalId.Parse(c.DataChannelID.LocalID),
                                        ShortId = c.DataChannelID.ShortID,
                                        NameObject = c.DataChannelID.NameObject is null
                                            ? null
                                            : new Domain.NameObject
                                            {
                                                NamingRule = c.DataChannelID.NameObject.NamingRule,
                                                CustomNameObjects = c.DataChannelID
                                                    .NameObject
                                                    .CustomProperties
                                                    ?.CopyProperties()
                                            }
                                    },
                                    Property = new Domain.Property
                                    {
                                        DataChannelType = new Domain.DataChannelType
                                        {
                                            Type = c.Property.DataChannelType.Type,
                                            UpdateCycle = c.Property.DataChannelType.UpdateCycle,
                                            CalculationPeriod = c.Property.DataChannelType.CalculationPeriod
                                        },
                                        Format = new Domain.Format
                                        {
                                            Type = c.Property.Format.Type,
                                            Restriction = c.Property.Format.Restriction is null
                                                ? null
                                                : new Domain.Restriction
                                                {
                                                    Enumeration = c.Property.Format.Restriction.Enumeration?.ToList(),
                                                    FractionDigits = (uint?)
                                                        c.Property.Format.Restriction.FractionDigits,
                                                    Length = (uint?)c.Property.Format.Restriction.Length,
                                                    MaxExclusive = c.Property.Format.Restriction.MaxExclusive,
                                                    MaxInclusive = c.Property.Format.Restriction.MaxInclusive,
                                                    MaxLength = (uint?)c.Property.Format.Restriction.MaxLength,
                                                    MinExclusive = c.Property.Format.Restriction.MinExclusive,
                                                    MinInclusive = c.Property.Format.Restriction.MinInclusive,
                                                    MinLength = (uint?)c.Property.Format.Restriction.MinLength,
                                                    Pattern = c.Property.Format.Restriction.Pattern,
                                                    TotalDigits = (uint?)c.Property.Format.Restriction.TotalDigits,
                                                    WhiteSpace = (Domain.WhiteSpace?)
                                                        c.Property.Format.Restriction.WhiteSpace
                                                }
                                        },
                                        Range = c.Property.Range is null
                                            ? null
                                            : new Domain.Range
                                            {
                                                Low = c.Property.Range.Low,
                                                High = c.Property.Range.High
                                            },
                                        Unit = c.Property.Unit is not null
                                            ? new Domain.Unit
                                            {
                                                UnitSymbol = c.Property.Unit.UnitSymbol,
                                                QuantityName = c.Property.Unit.QuantityName,
                                                CustomElements = c.Property.Unit.CustomProperties?.CopyProperties()
                                            }
                                            : null,
                                        QualityCoding = c.Property.QualityCoding,
                                        AlertPriority = c.Property.AlertPriority,
                                        Name = c.Property.Name,
                                        Remarks = c.Property.Remarks,
                                        CustomProperties = c.Property.CustomProperties?.CopyProperties()
                                    }
                                }
                        )
                        .ToList()
                }
            }
        };
    }
}

File: js/lib/transport/json/data-channel/Extensions.ts
import { LocalId } from "../../..";
import { DataChannelList, ShipId } from "../../domain";
import { Version } from "../../domain/data-channel/Version";
import { DataChannelListDto } from "./DataChannelList";

abstract class IExtension {
    public static toJsonDto: <Domain, Dto>(domain: Domain) => Dto;
    public static toDomainModel: <Domain, Dto>(dto: Dto) => Domain;
}

export class Extensions implements IExtension {
    public static toJsonDto(
        domain: DataChannelList.DataChannelListPackage
    ): DataChannelListDto.DataChannelListPackage {
        const p = domain.package;
        const h = domain.package.header;
        return {
            Package: {
                Header: {
                    Author: h.author,
                    DataChannelListID: {
                        ID: h.dataChannelListId.id,
                        TimeStamp: h.dataChannelListId.timestamp,
                        Version: h.dataChannelListId.version?.toString(),
                    },
                    ShipID: h.shipId.toString(),
                    DateCreated: h.dateCreated,
                    VersionInformation: h.versionInformation
                        ? {
                              NamingRule: h.versionInformation.namingRule,
                              NamingSchemeVersion:
                                  h.versionInformation.namingSchemeVersion,
                              ReferenceURL: h.versionInformation.referenceUrl,
                          }
                        : undefined,
                    ...h.customHeaders,
                },
                DataChannelList: {
                    DataChannel: p.dataChannelList.dataChannel.map(
                        DataChannelExtension.toJsonDto
                    ),
                },
            },
        };
    }

    public static async toDomainModel(
        dto: DataChannelListDto.DataChannelListPackage
    ): Promise<DataChannelList.DataChannelListPackage> {
        const dataChannels: DataChannelList.DataChannel[] = [];

        for (let dataChannel of dto.Package.DataChannelList.DataChannel) {
            dataChannels.push(await DataChannelExtension.toDomainModel(dataChannel));
        }

        const {
            DataChannelListID,
            DateCreated,
            ShipID,
            Author,
            VersionInformation,
            ...customHeaders
        } = dto.Package.Header;

        return {
            package: {
                header: {
                    dataChannelListId: {
                        id: DataChannelListID.ID,
                        timestamp: DataChannelListID.TimeStamp,
                        version: Version.parse(DataChannelListID.Version),
                    },
                    dateCreated: DateCreated,
                    shipId: ShipId.parse(ShipID),
                    author: Author,
                    versionInformation: VersionInformation
                        ? {
                              namingRule: VersionInformation.NamingRule,
                              namingSchemeVersion:
                                  VersionInformation.NamingSchemeVersion,
                              referenceUrl: VersionInformation.ReferenceURL,
                          }
                        : undefined,
                    customHeaders: customHeaders,
                },
                dataChannelList: {
                    dataChannel: dataChannels,
                },
            },
        };
    }
}

export class DataChannelExtension {
    public static toJsonDto(
        dataChannel: DataChannelList.DataChannel
    ): DataChannelListDto.DataChannel {
        return {
            DataChannelID: {
                LocalID: dataChannel.dataChannelId.localId.verboseMode
                    ? dataChannel.dataChannelId.localId.builder
                          .withoutVerboseMode()
                          .build()
                          .toString()
                    : dataChannel.dataChannelId.localId.toString(),
                ShortID: dataChannel.dataChannelId.shortId,
                NameObject: dataChannel.dataChannelId.nameObject
                    ? {
                          NamingRule: dataChannel.dataChannelId.nameObject.namingRule,
                          ...dataChannel.dataChannelId.nameObject.customProperties,
                      }
                    : undefined,
            },
            Property: {
                DataChannelType: {
                    Type: dataChannel.property.dataChannelType.type,
                    CalculationPeriod:
                        dataChannel.property.dataChannelType.calculationPeriod,
                    UpdateCycle: dataChannel.property.dataChannelType.updateCycle,
                },
                Format: {
                    Type: dataChannel.property.format.type,
                    Restriction: dataChannel.property.format.restriction
                        ? {
                              Enumeration:
                                  dataChannel.property.format.restriction.enumeration,
                              FractionDigits:
                                  dataChannel.property.format.restriction.fractionDigits,
                              Length: dataChannel.property.format.restriction.length,
                              MaxExclusive:
                                  dataChannel.property.format.restriction.maxExclusive,
                              MaxInclusive:
                                  dataChannel.property.format.restriction.maxInclusive,
                              MaxLength:
                                  dataChannel.property.format.restriction.maxLength,
                              MinExclusive:
                                  dataChannel.property.format.restriction.minExclusive,
                              MinInclusive:
                                  dataChannel.property.format.restriction.minInclusive,
                              MinLength:
                                  dataChannel.property.format.restriction.minLength,
                              Pattern: dataChannel.property.format.restriction.pattern,
                              TotalDigits:
                                  dataChannel.property.format.restriction.totalDigits,
                              WhiteSpace: dataChannel.property.format.restriction
                                  .whiteSpace
                                  ? (dataChannel.property.format.restriction
                                        .whiteSpace as unknown as DataChannelListDto.WhiteSpace)
                                  : undefined,
                          }
                        : undefined,
                },
                Range: dataChannel.property.range
                    ? {
                          Low: dataChannel.property.range.low,
                          High: dataChannel.property.range.high,
                      }
                    : undefined,
                Unit: dataChannel.property.unit
                    ? {
                          UnitSymbol: dataChannel.property.unit.unitSymbol,
                          QuantityName: dataChannel.property.unit.quantityName,
                          ...dataChannel.property.unit.customProperties,
                      }
                    : undefined,
                QualityCoding: dataChannel.property.qualityCoding,
                AlertPriority: dataChannel.property.alertPriority,
                Name: dataChannel.property.name,
                Remarks: dataChannel.property.remarks,
                ...dataChannel.property.customProperties,
            },
        };
    }

    public static async toDomainModel(
        dataChannel: DataChannelListDto.DataChannel
    ): Promise<DataChannelList.DataChannel> {
        const localId = await LocalId.parseAsync(
            dataChannel.DataChannelID.LocalID
        );

        const mapNameObject = (): DataChannelList.NameObject | undefined => {
            if (!dataChannel.DataChannelID.NameObject) return undefined;

            const { NamingRule, ...custom } =
                dataChannel.DataChannelID.NameObject;
            return {
                namingRule: NamingRule,
                customProperties: custom,
            };
        };

        const {
            DataChannelType,
            Format,
            Range,
            Unit,
            QualityCoding,
            AlertPriority,
            Name,
            Remarks,
            ...customProperties
        } = dataChannel.Property;

        const mapUnit = (): DataChannelList.Unit | undefined => {
            if (!Unit) return undefined;

            const { UnitSymbol, QuantityName, ...custom } = Unit;
            return {
                unitSymbol: Unit.UnitSymbol,
                quantityName: Unit.QuantityName,
                customProperties: custom,
            };
        };

        return {
            dataChannelId: {
                localId: localId,
                shortId: dataChannel.DataChannelID.ShortID,
                nameObject: mapNameObject(),
            },
            property: {
                dataChannelType: {
                    type: DataChannelType.Type,
                    calculationPeriod: DataChannelType.CalculationPeriod,
                    updateCycle: DataChannelType.UpdateCycle,
                },
                format: {
                    type: Format.Type,
                    restriction: Format.Restriction
                        ? {
                              enumeration: Format.Restriction.Enumeration,
                              fractionDigits: Format.Restriction.FractionDigits,
                              length: Format.Restriction.Length,
                              maxExclusive: Format.Restriction.MaxExclusive,
                              maxInclusive: Format.Restriction.MaxInclusive,
                              maxLength: Format.Restriction?.MaxLength,
                              minExclusive: Format.Restriction.MinExclusive,
                              minInclusive: Format.Restriction.MinInclusive,
                              minLength: Format.Restriction.MinLength,
                              pattern: Format.Restriction.Pattern,
                              totalDigits: Format.Restriction.TotalDigits,
                              whiteSpace: Format.Restriction.WhiteSpace,
                          }
                        : undefined,
                },
                range: Range
                    ? {
                          low: Range.Low,
                          high: Range.High,
                      }
                    : undefined,
                unit: mapUnit(),
                qualityCoding: QualityCoding,
                alertPriority: AlertPriority,
                name: Name,
                remarks: Remarks,
                customProperties: customProperties,
            },
        };
    }
}
